File: nlpcraft/src/main/scala/org/apache/nlpcraft/NCResult.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.nlpcraft;

import org.apache.nlpcraft.internal.util.*;

import java.io.Serializable;
import java.util.Collection;

/**
 *
 */
public class NCResult implements Serializable {
    /** Data Model result text. */
    private Object body;

    /** Data Model result type. */
    private NCResultType type;

    /** ID of the intent. */
    private String intentId;

    /**
     * Creates new result with given body and type.
     *
     * @param body Result body.
     * @param type Result type.
     */
    public NCResult(Object body, NCResultType type) {
        assert body != null;
        assert type != null;

        this.body = body;
        this.type = type;
    }

    /**
     * No-arg constructor.
     */
    public NCResult() {
        // No-op.
    }

    /**
     * Sets result body.
     *
     * @param body Result body.
     */
    public void setBody(Object body) {
        this.body = body;
    }

    /**
     * Set result type.
     *
     * @param type Result type.
     */
    public void setType(NCResultType type) {
        this.type = type;
    }

    /**
     * Gets result type.
     *
     * @return Result type.
     */
    public NCResultType getType() {
        return type;
    }

    /**
     * Gets result body.
     *
     * @return Result body.
     */
    public Object getBody() {
        return body;
    }

    /**
     * Get optional intent ID.
     *
     * @return Intent ID or {@code null} if intent ID is not available.
     */
    public String getIntentId() {
        return intentId;
    }

    /**
     * Sets optional intent ID.
     *
     * @param intentId Intent ID to set for this result.
     */
    public void setIntentId(String intentId) {
        this.intentId = intentId;
    }
}

File: nlpcraft/src/main/scala/org/apache/nlpcraft/NCResult.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.nlpcraft

/**
  * Convenient factory for creating [[NCResult]] instances.
  */
object NCResult:
    /**
      * Creates new result instance with given parameters.
      *
      * @param body Result body.
      * @param resultType Result type.
      * @param intentId Intent ID that produced the result.
      */
    def apply(body: Any, resultType: NCResultType, intentId: String): NCResult =
        new NCResult():
            def getBody: Any = body
            def getType: NCResultType = resultType
            def getIntentId: Option[String] = intentId.?

    /**
      * Creates new result instance with given parameters. Note that intent ID will be set to `None`.
      *
      * @param body Result body.
      * @param resultType Result type.
      */
    def apply(body: Any, resultType: NCResultType): NCResult =
        new NCResult():
            def getBody: Any = body
            def getType: NCResultType = resultType
            def getIntentId: Option[String] = None

    /**
      * Creates new result instance with given parameters. Note that intent ID will be set to `None` and
      * result type will be set to [[NCResultType.ASK_RESULT]].
      *
      * @param body Result body.
      */
    def apply(body: Any): NCResult =
        new NCResult() :
            def getBody: Any = body
            def getType: NCResultType = NCResultType.ASK_RESULT
            def getIntentId: Option[String] = None

/**
  * Intent callback result.
  */
trait NCResult:
    /**
      * Body of the result.
      */
    def getBody: Any

    /**
      * Type of the result.
      */
    def getType: NCResultType

    /**
      * Optional ID of the intent, if any, that produced that result.
      */
    def getIntentId: Option[String]